internals/overlord/servstate/handlers.go

```diff
@@ -338,17 +355,31 @@ func (s *serviceData) startInternal() error {
 	s.cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
 
 	// Copy environment to avoid updating original.
-	environment := make(map[string]string)
-	for k, v := range s.config.Environment {
-		environment[k] = v
+	var environment map[string]string
+	if s.workload != nil && len(s.workload.Environment) > 0 {
+		environment = maps.Clone(s.workload.Environment)
+	} else if len(s.config.Environment) > 0 {
+		environment = maps.Clone(s.config.Environment)
+	} else {
+		environment = make(map[string]string)
 	}
 
 	s.cmd.Dir = s.config.WorkingDir
 
 	// Start as another user if specified in plan.
-	uid, gid, err := osutil.NormalizeUidGid(s.config.UserID, s.config.GroupID, s.config.User, s.config.Group)
-	if err != nil {
-		return err
+	var uid, gid *int
+	if s.config.UserID != nil || s.config.GroupID != nil || s.config.User != "" || s.config.Group != "" {
+		// User/group config from the service takes precedence if any of them are set
```

Comment:
Do we have an explicit specification about mixed cases? 

As in, what should happen if the service config specifies: 1) a user and 2) a workload but 3) no group? As a user I think I would expect that the service would then use the group from its workload and the user from the service config. In the code as it is, in such a case the user written in the service config would be silently ignored.